Implement custom AnnotationSchema subclass for NamedTuple

gwax opened the pull request:
I started building out something very similar with a focus on supporting `typing.NamedTuple` as a central model construct. After stumbling on your project, it seemed easier to port my work to live on top of `marshmallow-annotations` than to try and roll my own parallel project.

justanr commented:
👍 I'll have to take a closer look when I'm home from work but first glance looks good. Caught a few documentation errors too, appreciate that.

justanr commented on src/marshmallow_annotations/namedtuple.py:
```diff
@@ -0,0 +1,68 @@
+        """Post dump, strip default fields from serialized output."""
+        return {
+            k: v for k, v in data.items() if
+            v != self.opts.target._field_defaults.get(k)
```
What's the motivation for stripping these values? Seems like they should be included to me.

gwax replied:
The motivation is two-fold:

1. Allow round tripping of values that take advantage of missing optional fields.
2. Minimize json payload size.

Without this, we have a situation where things don't round trip properly:

```python
class Something(NamedTuple):
    a: int
    b: Optional[int]
    c: int

class SomethingSchema(NamedTupleSchema):
    class Meta:
        target = Something

schema = SomethingSchema()

input = {'a': 1, 'c': 3}
schema.dump(
    schema.load(input).data
).data
# current: {'a': 1, 'c': 3}
# without stripping: {'a': 1, 'b': None, 'c': 3}
```

gwax replied:
That said, I've been having mixed feelings about this myself and might go back and put it behind an Opts flag.

justanr replied:
Opts flag is what I'd expect if value stripping ng was left in.

justanr commented on src/marshmallow_annotations/namedtuple.py:
```diff
@@ -0,0 +1,68 @@
+
+    def _process(self, meta, schema):
+        self._extract_from_parents(schema, self._extract_from)
+        if self.converter_factory is BaseConverter:
```
How about `not issubclass(self.converter_factory, NamedTupleConverter)` instead. If cover other converters such as attrs or dataclass

justanr commented:
Being curious, how easy was this to implement? Is there anything that the library should do differently in your opinion?

gwax commented:
This was pretty easy to implement because it mostly layered on top of what you had already done.

The converter portion took a little thinking but was totally reasonable.

Looking at #11, I think it will be pretty easy to abstract a few pieces of this code and add support for the Python 3.6 [dataclasses backport](https://github.com/ericvsmith/dataclasses). In anticipation of that, I might rearrange this and move `marshmallow_annotations.namedtuple` to `marshmallow_annotations.ext.namedtuple`.

gwax commented:
I have moved the core default handling functionality into BaseConverter, put default value stripping behind a `Meta` option flag, and moved the namedtuple code to `marshmallow_annotations.ext`.

I think this is good to go now.

justanr commented on src/marshmallow_annotations/converter.py:
```diff
@@ -49,7 +49,11 @@ class BaseConverter(AbstractConverter):
+        self,
+        typehint: type,
+        opts: ConfigOptions = None
+    ) -> marshmallow.fields.FieldABC:
```
Not gonna block on this, but I tend to leave things how black formats them, in this case it's 

      def convert(
          self, typehint: type, opts: ConfigOptions = None
      ) -> marshmallow.fields.FieldABC:

justanr commented on src/marshmallow_annotations/converter.py:
```diff
@@ -13,16 +13,16 @@
+        hasattr(typehint, '__origin__') and
+        typehint.__origin__ is Union and
+        len(typehint.__args__) == 2 and
+        NoneType in typehint.__args__
```
`and` on the beginning of the line with the `# noqa` tags, flake8 is actually wrong in this case. and double quotes for consistency.

justanr replied:
Hm. I wish the docs were better about interacting with typehints and/or there were facilities builtin to help with this sort of code generation. With dataclasses in 3.7 I'm sure the core devs felt this pain too.

justanr commented on src/marshmallow_annotations/converter.py:
```diff
@@ -77,11 +83,13 @@ def _field_from_typehint(self, typehint, kwargs=None):
             required = False
-            typehint = typehint.__args__[0]
+            missing = None
+            [typehint] = [t for t in typehint.__args__ if t is not NoneType]
```
eh, I think it's more readable before but I think this is more correct.

justanr commented on src/marshmallow_annotations/ext/namedtuple.py:
```diff
@@ -0,0 +1,41 @@
+
+from marshmallow_annotations.scheme import AnnotationSchema, AnnotationSchemaOpts
+
+
```
Add `__all__ = ("NamedTupleSchemaOpts", "NamedTupleSchema")`

justanr commented on docs/extensions.rst:
```diff
@@ -0,0 +1,60 @@
+
+Additionally, the ``Meta`` class provides you with the option flag
+``dump_default_args`` to control whether attribute values matching defaults
+should be dumped or ignored::
```
- should be dumped or ignored::
    + should be dumped or ignored, the default is to dump them::

Additionally, and I won't block on this, I'm a big fan of using two-value enums rather than booleans when possible:

    class DefaultArgBehavior(Enum):
        dump = auto()
        ignore = auto()

And using `default_args = DefaultArgBehavior.dump`

But that's purely personal preference.

justanr commented on docs/conf.py:
```diff
@@ -167,6 +167,6 @@
 # Example configuration for intersphinx: refer to the Python standard library.
 intersphinx_mapping = {
-    'python': ('https://docs.python.org/', None),
+    'python': ('https://docs.python.org/3', None)
```
I left the 3 off on purpose. This is okay since it'll be tied to whatever release this goes in and Python 3.7 will be the most current, but I'd prefer to always point this at latest.

justanr commented on docs/extensions.rst:
```diff
@@ -0,0 +1,60 @@
+Specialized extensions are provided for specific use cases.
+
+
+*****************
```
Since supporting dataclasses and attrs is a stated goal, I'd rather break this into sub-pages rather than doing one huge page from the get go. Less likely to break links in the future.

justanr commented on src/marshmallow_annotations/converter.py:
```diff
@@ -60,6 +64,8 @@ def convert_all(
     ) -> GeneratedFields:
         configs = configs if configs is not None else {}
+        for k, default in self._get_field_defaults(target).items():
+            configs[k] = {'missing': default, **configs.get(k, {})}
```
double quotes.

justanr commented on src/marshmallow_annotations/converter.py:
```diff
@@ -100,3 +109,10 @@ def _get_type_hints(self, item):
+
+    def _get_field_defaults(self, item):
+        """Retrieve default values if item is a namedtuple."""
+        if hasattr(item, '_field_defaults'):
```
double quotes.

justanr replied:
Additionally, I don't think this belongs in the base:

```python
class SomeThing:
    a: int = 1

hasattr(SomeThing, '_field_defaults') # False
```

If we do put it here, the default implementation should be:

```python
def _get_field_defaults(self, target):
    return {}
```

And then each extension can provide their own implementation if needed. We'd have to document it as a non-public override point but that's not uncommon even in stricter OO languages.

justanr commented on src/marshmallow_annotations/ext/namedtuple.py:
```diff
@@ -0,0 +1,41 @@
+        if self.opts.dump_default_fields:
+            return data
+        else:
+            default_values = self.opts.converter._get_field_defaults(self.opts.target)
```
Could we store the defaults on the opts so we don't need to look them up again? Probably not a big issue but a thought. The downside would be if a default is mutable, though you have other issues in that case as well.

gwax replied:
In this case, I think that's unnecessary complexity. Under the hood this call becomes a field lookup of `item._field_defaults` on the target namedtuple. This ought to have nearly equivalent performance to storing the values on opts instead.

gwax commented:
I believe that I have addressed your feedback.

justanr commented:
:+1:

justanr commented:
:beers: Just released 2.1.0 with this. Figured you wanted to use it in something so might as well make it easy to use.
